File: 2576-minimum-penalty-for-a-shop/2576-minimum-penalty-for-a-shop.go
func bestClosingTime(customers string) int {
    close_hour := 0
    min_penalty := 0
    curr_penalty := 0
    for i,item := range(customers){
        if string(item) == "Y"{
            curr_penalty-=1
            if curr_penalty < min_penalty{
                min_penalty = curr_penalty
                close_hour = i+1
            }
        }else{
            curr_penalty+=1
        }
    }
    return close_hour
}

File: 2576-minimum-penalty-for-a-shop/2576-minimum-penalty-for-a-shop.ts
function bestClosingTime(customers: string): number {

    let ans :number = 0;
    let profit: number = 0;
    let maxProfit: number = 0;

    for (let i = 0; i < customers.length; i++) {
      profit += customers[i] === 'Y' ? 1 : -1;
      if (profit > maxProfit) {
        maxProfit = profit;
        ans = i + 1;
      }
    }

 return ans;
};